go/src/koding/kites/kloud/provider/do/stack.go

```diff
@@ -0,0 +1,244 @@
+// creates a ssh key on behalf of Koding, that will be later used during
+// ApplyTemplate()
+func (s *Stack) BootstrapTemplates(c *stack.Credential) ([]*stack.Template, error) {
+	type tmplData struct {
+		KeyName   string
+		PublicKey string
+	}
+
+	// fill the template
+	var buf bytes.Buffer
+	if err := bootstrapTmpl.Execute(&buf, &tmplData{
+		KeyName:   s.keyName(),
+		PublicKey: s.Keys.PublicKey,
+	}); err != nil {
+		return nil, err
+	}
+
+	return []*stack.Template{
+		{Content: buf.String()},
+	}, nil
+}
+
+// keyName returns the keyName used for the bootstrap data
+func (s *Stack) keyName() string {
+	return fmt.Sprintf("koding-deployment-%s-%s", s.Req.Username, s.BootstrapArg().GroupName)
+}
+
+// ApplyTemplate enhances and updates the DigitalOcean terraform template. It
+// updates the various sections of the template, such as Provider, Resources,
+// Variables, etc... so it can be executed without any problems
+func (s *Stack) ApplyTemplate(c *stack.Credential) (*stack.Template, error) {
+	cred, ok := c.Credential.(*Credential)
+	if !ok {
+		return nil, fmt.Errorf("credential is not of type do.Credential: %T", c.Credential)
+	}
+
+	bootstrap, ok := c.Bootstrap.(*Bootstrap)
+	if !ok {
+		return nil, fmt.Errorf("bootstrap is not of type do.Bootstrap: %T", c.Bootstrap)
+	}
+
+	template := s.Builder.Template
+	template.Provider["digitalocean"] = map[string]interface{}{
+		"token": cred.AccessToken,
+	}
+
+	keyID, err := strconv.Atoi(bootstrap.KeyID)
+	if err != nil {
+		return nil, err
+	}
+
+	droplet, err := s.modifyDroplets(keyID)
+	if err != nil {
+		return nil, err
+	}
+
+	template.Resource["digitalocean_droplet"] = droplet
+
+	if err := template.Flush(); err != nil {
+		return nil, err
+	}
+
+	if err := template.ShadowVariables("FORBIDDEN", "do_token"); err != nil {
+		return nil, err
+	}
+
+	content, err := template.JsonOutput()
+	if err != nil {
+		return nil, err
+	}
+
+	return &stack.Template{
+		Content: content,
+	}, nil
+}
+
+// modifyDroplets returns a modified 'digitalocean_droplet' terraform resource
+// from the stack that changes things like image, injects kite and ssh_key,
+// etc...
+func (s *Stack) modifyDroplets(keyID int) (map[string]map[string]interface{}, error) {
+	var resource struct {
+		Droplet map[string]map[string]interface{} `hcl:"digitalocean_droplet"`
+	}
+
+	if err := s.Builder.Template.DecodeResource(&resource); err != nil {
+		return nil, err
+	}
+
+	if len(resource.Droplet) == 0 {
+		return nil, errors.New("there are no droplets available")
+	}
+
+	// we might have multipel droplets, iterate over all of them
+	for dropletName, droplet := range resource.Droplet {
+		// Do not overwrite SSH key pair with the bootstrap one
+		// when user sets it explicitly in a template.
+		if s, ok := droplet["ssh_keys"]; !ok {
+			droplet["ssh_keys"] = []int{keyID}
+		} else if keyIds, ok := s.([]int); ok {
+			keys := []int{keyID}
+			if len(keyIds) != 0 {
```

Comment:
even `var keyIds []int = nil` can be expanded `keyIds...` so no need for this check.